pkg/querier/queryrange/log_result_cache.go

```diff
@@ -0,0 +1,280 @@
+// Currently it only caches empty filter queries, this is because those are usually easily and freely cacheable.
+// Log hits are difficult to handle because of the limit query parameter and the size of the response.
+// In the future it could be extended to cache non-empty query results.
+// see https://docs.google.com/document/d/1_mACOpxdWZ5K0cIedaja5gzMbv-m0lUVazqZd2O4mEU/edit
+func NewLogResultCache(logger log.Logger, limits Limits, cache cache.Cache, shouldCache queryrangebase.ShouldCacheFn, metrics *LogResultCacheMetrics) queryrangebase.Middleware {
+	if metrics == nil {
+		metrics = NewLogResultCacheMetrics(nil)
+	}
+	return queryrangebase.MiddlewareFunc(func(next queryrangebase.Handler) queryrangebase.Handler {
+		return &logResultCache{
+			next:        next,
+			limits:      limits,
+			cache:       cache,
+			logger:      logger,
+			shouldCache: shouldCache,
+			metrics:     metrics,
+		}
+	})
+}
+
+type logResultCache struct {
+	next        queryrangebase.Handler
+	limits      Limits
+	cache       cache.Cache
+	shouldCache queryrangebase.ShouldCacheFn
+
+	metrics *LogResultCacheMetrics
+	logger  log.Logger
+}
+
+func (l *logResultCache) Do(ctx context.Context, req queryrangebase.Request) (queryrangebase.Response, error) {
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+
+	if l.shouldCache != nil && !l.shouldCache(req) {
+		return l.next.Do(ctx, req)
+	}
+
+	maxCacheFreshness := validation.MaxDurationPerTenant(tenantIDs, l.limits.MaxCacheFreshness)
+	maxCacheTime := int64(model.Now().Add(-maxCacheFreshness))
+	if req.GetStart() > maxCacheTime {
+		return l.next.Do(ctx, req)
+	}
+
+	lokiReq, ok := req.(*LokiRequest)
+	if !ok {
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "invalid request type %T", req)
+	}
+
+	interval := validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, l.limits.QuerySplitDuration)
+	// skip caching by if interval is unset
+	if interval == 0 {
+		return l.next.Do(ctx, req)
+	}
+	// The first subquery might not be aligned.
+	alignedStart := time.Unix(0, lokiReq.GetStartTs().UnixNano()-(lokiReq.GetStartTs().UnixNano()%interval.Nanoseconds()))
+	// generate the cache key based on query, tenant and start time.
+	cacheKey := fmt.Sprintf("log:%s:%s:%d", tenant.JoinTenantIDs(tenantIDs), req.GetQuery(), alignedStart.UnixNano()/(interval.Nanoseconds()))
+
+	_, buff, _, err := l.cache.Fetch(ctx, []string{cache.HashKey(cacheKey)})
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error fetching cache", "err", err, "cacheKey", cacheKey)
+		return l.next.Do(ctx, req)
+	}
+	// we expect only one key to be found or missing.
+	if len(buff) > 1 {
+		level.Warn(l.logger).Log("msg", "unexpected length of cache return values", "buff", len(buff))
+		return l.next.Do(ctx, req)
+	}
+
+	if len(buff) == 0 {
+		// cache miss
+		return l.handleMiss(ctx, cacheKey, lokiReq)
+	}
+
+	// cache hit
+	var cachedResquest LokiRequest
```

Comment:
here's another `cachedResquest`. I see others have commented that this should probably be `cachedRequest`?

Reply:
it's already gone :)